src/peer.cpp

```diff
@@ -238,20 +224,79 @@ void PeerConnection::handle_response(const asio::error_code& err, std::size_t) {
     //       mentions that maybe socket.async_read_some is better to read > 512
     //       bytes at a time
     asio::async_read(
-        socket_, response_, asio::transfer_at_least(1),
+        *socket_, response_, asio::transfer_at_least(1),
         [this](const auto& ec, auto s) { handle_response(ec, s); });
   } else if (err != asio::error::eof) {
     logger()->error("Response failed: {}", err.message());
   } else {
     logger()->debug("handle_response EOF");
-    peer_.torrent().disconnected(&peer_);
+    peer_.disconnected();
   }
 }
 
 void PeerConnection::stop() {
   resolver_.cancel();
-  acceptor_.close();
-  socket_.close();
+  socket_->close();
+}
+
+// NOLINTNEXTLINE(cppcoreguidelines-avoid-non-const-global-variables)
+std::map<ListeningPort, PeerAcceptor> PeerAcceptor::m_acceptors;
+
+PeerAcceptor::PeerAcceptor(ListeningPort port,
+                           asio::io_context& io_context,
+                           std::string bind_address)
+    : m_port(port), m_io_context(io_context), m_bind_address(std::move(bind_address)) {
+  logger()->trace(PRETTY_FUNCTION);
+  // For now rethrowing - can consider asio::detached later?
+  co_spawn(m_io_context, listen(), [](auto e) { std::rethrow_exception(e); });
+}
+
+asio::awaitable<void> PeerAcceptor::listen() {
+  logger()->trace(PRETTY_FUNCTION);
+  // FIXME: We need to be able to bind to a different IP than the default
+  // localhost
+  //        For testing we can setup a new local ip with
+  //        sudo ip addr add 192.168.1.10/32 dev lo
+
+  asio::ip::tcp::acceptor acceptor{m_io_context, tcp::v4()};
+  const asio::socket_base::reuse_address option(true);
+  acceptor.set_option(option);
+  acceptor.bind(asio::ip::tcp::endpoint(asio::ip::make_address(m_bind_address),
+                                        m_port.get()));
+  acceptor.listen();
+
+  logger()->info("Listening for incoming connections on {}:{}",
+                 acceptor.local_endpoint().address().to_string(), m_port.get());
+
+  while (true) {
+    try {
+      auto socket = co_await acceptor.async_accept(asio::use_awaitable);
+      auto ip = socket.remote_endpoint().address().to_string();
+      auto port = socket.remote_endpoint().port();
+      logger()->info("Accepted new connection from {}:{}", ip, port);
+
+      // Read data from socket
+      bytes buffer;
+      // TODO: Check how timeout for this works - does it wait forever?
```

Comment:
TODO: Check how timeout for this works - does it wait forever?

Reply:
From what the doc it seem to be ok - can revisit if it turns out to be a problem later.